Userland/Libraries/LibWeb/HTML/Worker.cpp

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (c) 2022, Ben Abraham <ben.d.abraham@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+#include <LibJS/Runtime/Realm.h>
+#include <LibWeb/Bindings/MainThreadVM.h>
+#include <LibWeb/Bindings/WorkerWrapper.h>
+#include <LibWeb/DOM/ExceptionOr.h>
+#include <LibWeb/HTML/Scripting/Environments.h>
+#include <LibWeb/HTML/Worker.h>
+
+namespace Web::HTML {
+
+// https://html.spec.whatwg.org/multipage/workers.html#dedicated-workers-and-the-worker-interface
+/*
+Worker::Worker(String&, WorkerOptions const&)
+    : Worker(script_url, options)
+{
+}
+*/
+
+Worker::Worker(String& script_url, WorkerOptions const& options, DOM::Document& document)
+    : m_script_url(script_url)
+    , m_options(options)
+    , m_document(&document)
+    , m_custom_data()
+    , m_worker_vm(JS::VM::create(adopt_own(m_custom_data)))
+    , m_interpreter(JS::Interpreter::create<JS::GlobalObject>(m_worker_vm))
+    , m_interpreter_scope(*m_interpreter)
+    , m_execution_context(m_worker_vm->heap())
+    , m_implicit_port(MessagePort::create())
+{
+}
+
+DOM::ExceptionOr<NonnullRefPtr<Worker>> Worker::create(String& script_url, WorkerOptions const& options, DOM::Document& document)
+{
+    dbgln_if(WEB_WORKER_DEBUG, "Worker: Create={}", script_url);
+
+    // Returns a new Worker object. scriptURL will be fetched and executed in the background,
+    // creating a new global environment for which worker represents the communication channel.
+    // options can be used to define the name of that global environment via the name option,
+    // primarily for debugging purposes. It can also ensure this new global environment supports
+    // JavaScript modules (specify type: "module"), and if that is specified, can also be used
+    // to specify how scriptURL is fetched through the credentials option.
+
+    // FIXME: 1. The user agent may throw a "SecurityError" DOMException if the request violates
+    // a policy decision (e.g. if the user agent is configured to not allow the page to start dedicated workers).
+    // Technically not a fixme if our policy is not to throw errors :^)
+
+    // 2. Let outside settings be the current settings object.
+    auto& outside_settings = document.relevant_settings_object();
+
+    // FIXME: 3. Parse the scriptURL argument relative to outside settings.
+    auto url = document.parse_url(script_url);
+
+    // 4. If this fails, throw a "SyntaxError" DOMException.
+    if (!url.is_valid()) {
+        dbgln_if(WEB_WORKER_DEBUG, "WebWorker: Invalid URL loaded '{}'.", script_url);
+        return DOM::SyntaxError::create("Failed to parse selector");
+    }
+
+    // 5. Let worker URL be the resulting URL record.
+    auto worker_url = url;
```

Comment:
You can skip steps like this, perhaps adding a comment (although I don't think that's needed here)